addons/sale_coupon/tests/test_program_numbers.py

```diff
@@ -1082,3 +1082,90 @@ def test_program_free_prods_with_min_qty_and_reward_qty_and_rule(self):
 
         self.assertEqual(order.amount_total, 500.0, 'The price must be 500.0 since two conference chairs are free and the user only bought one')
         self.assertEqual(order.order_line[2].price_total, -100.0, 'The last order line should apply a reduction of 100.0 since there is one conference chair that cost 100.0')
+
+    def test_program_free_product_different_than_rule_product_with_multiple_application(self):
+        order = self.empty_order
+        product_D = self.env['product.product'].create({
+            'name': 'Product D',
+            'list_price': 100,
+            'sale_ok': True,
+            'taxes_id': [(6, 0, [])],
+
+        })
+        product_E = self.env['product.product'].create({
+            'name': 'Product E',
+            'list_price': 5,
+            'sale_ok': True,
+            'taxes_id': [(6, 0, [])],
+
+        })
+        immediate_program = self.env['sale.coupon.program'].create({
+            'name': '3 Product D = 1 free Product E',
+            'promo_code_usage': 'no_code_needed',
+            'reward_type': 'product',
+            'reward_product_id': product_E.id,
+            'rule_products_domain': "[('id', 'in', [%s])]" % (product_D.id),
+            'active': True,
+        })
+
+        sol_D = self.env['sale.order.line'].create({
+            'product_id': product_D.id,
+            'product_uom_qty': 6.0,
+            'order_id': order.id,
+        })
+        sol_E = self.env['sale.order.line'].create({
+            'product_id': product_E.id,
+            'product_uom_qty': 1.0,
+            'order_id': order.id,
+        })
+
+        order.recompute_coupon_lines()
+
+        self.assertEqual(len(order.order_line), 3, 'The order must contain 3 order lines including one for a free Product E')
+        self.assertEqual(order.amount_total, 600.0, 'The price must be 600.0 since the Product E is free')
```

Comment:
```suggestion
        self.assertEqual(order.amount_total, 600.0, "The price must be 600.0 since the Product E is free: +600.00 (product D) + 5.00 (product E) - 5.00 (free product E)")
```